tagger/plot/basic.py

```diff
@@ -500,6 +513,83 @@ def plot_shaply(model, X_test, class_labels, input_vars, plot_dir):
         plt.savefig(plot_dir+"/shap_summary_reg.pdf",bbox_inches='tight')
         plt.savefig(plot_dir+"/shap_summary_reg.png",bbox_inches='tight')
 
+def efficiency(y_pred, y_test, reco_pt_test, class_labels, plot_dir):
+    
+    save_dir = os.path.join(plot_dir, 'efficiencies')
+    os.makedirs(save_dir, exist_ok=True)
+
+    # pT coordinate points for plotting
+    pt_points = [np.mean((PT_BINS[i], PT_BINS[i + 1])) for i in range(len(PT_BINS) - 1)]
+
+    def plot_efficiency(df_wp_loose, df_wp_medium, df_wp_tight, plot_name):
+        # Plot the efficiency
+        fig = plt.figure(figsize=style.FIGURE_SIZE)
+
+        ax = df_wp_loose.plot.line(x="midpoints", y="wp_loose", yerr="err(wp_loose)", label = "Loose (mistag = 50%)", color="blue", linestyle='-')
+        df_wp_medium.plot.line(x="midpoints", y="wp_medium", yerr="err(wp_medium)", label = "Medium (eff = 20%)", ax = ax, color="orange", linestyle='-')
+        df_wp_tight.plot.line(x="midpoints", y="wp_tight", yerr="err(wp_tight)", label = "Tight (eff = 10%)", ax = ax, color="green", linestyle='-')
+        plt.xlabel(r'Jet $p_T$ [GeV]')
+        plt.ylabel('Tagging efficiency')
+        plt.ylim(0., 1.3)
+        # plt.xlim(0., 1000.)
+        plt.legend()
+
+        hep.cms.label(llabel=style.CMSHEADER_LEFT,rlabel=style.CMSHEADER_RIGHT,ax=ax,fontsize=style.CMSHEADER_SIZE)
+
+        # Save the plot
+        save_path = os.path.join(save_dir, plot_name)
+        plt.savefig(f"{save_path}.pdf", bbox_inches='tight')
+        plt.savefig(f"{save_path}.png", bbox_inches='tight')
+        plt.cla()
+        plt.close()
+
+    eff_selection = {
+        'taus': [class_labels['taup'], class_labels['taum']],
+        'b': [class_labels['b']],
+    }
+
+    for key in eff_selection.keys():
+        selection = sum(y_test[:, idx] for idx in eff_selection[key]) > 0
+        summed_y_true = sum(y_test[:, idx] for idx in eff_selection[key])
+        summed_y_score = sum(y_pred[:, idx] for idx in eff_selection[key])
+
+        # Compute FPR, TPR, and AUC
+        fpr, tpr, thres = roc_curve(summed_y_true, summed_y_score)
+        roc_auc = auc(fpr, tpr)
+
+        # get working points for 50%, 20%, 10%
+        wp_loose = thres[np.argmin(np.abs(np.array(fpr)-0.5))]
+        wp_medium = thres[np.argmin(np.abs(np.array(fpr)-0.2))]
+        wp_tight = thres[np.argmin(np.abs(np.array(fpr)-0.1))]
+
+        # some tricks to get proper uncertainties in the efficiencies
+        data_eff = ak.to_dataframe(ak.Array({"truth": summed_y_true[selection], "pt": reco_pt_test[selection], "pred": summed_y_score[selection]}))
+        data_eff["wp_loose"] = data_eff["pred"] > wp_loose
+        data_eff["wp_medium"] = data_eff["pred"] > wp_medium
+        data_eff["wp_tight"] = data_eff["pred"] > wp_tight
+        import collections
```

Comment:
Could these imports be brought in at the start of the file rather than inline

Reply:
Sure, can update it. I saw similar imports in other files, should I move them all there, too?